erf and exp/trig integral functions: rewrite as integrals

cbm755 opened the pull request:
Many of these funtions are actually defined in terms of integrals
or at least can be written that way.  Now this will work:
`erf(z).rewrite('Integral')`.  Similarly for the other error fcns
as well as exponential integral fcns and trig integral fcns.

cbm755 committed:
erf and exp/trig integral functions: rewrite as integrals

Many of these funtions are actually defined in terms of integrals
or at least can be written that way.  Now this will work:
`erf(z).rewrite('Integral')`.  Similarly for the other error fcns
as well as exponential integral fcns and trig integral fcns.

cbm755 commented:
Please review.  But note I just copied how `transforms.py` did it without thinking very much.  If a different design is preferable, let me know.

cbm755 commented:
Questions: is `t = Symbol('t', Dummy=True)` ok or should it be `t = Dummy('t')`?  Former does look nicer for my tests.  But OTOH the latter renders as `_t`, which gives some indication a dummy variable was introduced.

cbm755 committed:
Use correct integral formula for expint

Previous implementation of `as_integeral` blindly copied the
documentation, which had a typo.  Also fixed typo, updated tests.

cbm755 commented:
@kritkaran94 @sampadsaha5 both of you seem to have been interested in `rewrite` in the past year or so.  Can you review this PR?

(I'd like to do this change for elliptic functions too but I'd like to know its the right approach first.)

jksuom commented:
> Questions: is t = Symbol('t', Dummy=True) ok or should it be t = Dummy('t')? 

I also think that `t` looks nicer than `_t`, but it seems that `Dummy=True` will not create a Dummy but an ordinary Symbol having 'Dummy' among its assumptions.
```
>>> t = Symbol('t', Dummy=True)
>>> t.is_Dummy
False
>>> t.assumptions0['Dummy']
True
```
Perhaps the interface could be extended to admit a user defined integration variable. Dummies are sometimes inconvenient to deal with.

cbm755 committed:
error fcns: use Dummy symbols when rewriting as Integrals

cbm755 commented:
Oh thanks!  `Symbol('t', Dummy=True)` seems like a bad idea then.

I'm not sure how the we might extend it to use a user defined integration variable: I'm not sure it makes sense to associate it with the `elliptic_e` object itself (although the `IntegralTransforms` seem to do this!)

For now, I'll use a regular Dummy and use `str()` in the tests.

cbm755 commented:
Hmmm, now I get pickling test failures, because (I think) the new `as_integral` properties of `erf` makes a Dummy.  And two integrals with different variables of integration are not `==` (#5539).

> Dummies are sometimes inconvenient to deal with.

@jksuom Yes I'm starting to agree!

Here's the test failure I'm getting.  Reminds me of #11847.
````
____________ sympy/utilities/tests/test_pickling.py:test_functions _____________

  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/sympy/utilities/tests/test_pickling.py", line 213, in test_functions

    check(c)

  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/sympy/utilities/tests/test_pickling.py", line 76, in check

    c(a, b, d1)

  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/sympy/utilities/tests/test_pickling.py", line 75, in c

    assert getattr(b, i) == attr, "%s != %s" % (getattr(b, i), attr)

AssertionError: 2*Integral(exp(-_t**2), (_t, 0, x))/sqrt(pi) != 2*Integral(exp(-_t**2), (_t, 0, x))/sqrt(pi)
````

cbm755 commented:
Hmmm, and wtf does #12136 pass that pickling test (#12136 tries this same approach for `elliptic_f` et al)....  Ahh, its because `erf` is in a list in `test_pickling:test_functions()` but `elliptic_f` is not.

'tis simple: just remove `erf` from the pickling test...  please may I do that?

cbm755 committed:
tests: drop erf from pickling test

Well... because it makes integrals from its "as_integral" property
and those integrals have a Dummy variable of integration.  When they
are unpickled, the "as_integral" properties are different (because its
a new Dummy variable).  Related to #5539 whereby integrals with same
integrands and limits but different variables of integration are not
equal.

jksuom commented:
If `as_integral` were an ordinary method, it would be possible to add an argument for the variable. It is defined as a property in some integral transforms, but I think that could be changed. It seems that most of the other `as_...` type methods admit arguments.

cbm755 committed:
Revert "tests: drop erf from pickling test"

This reverts commit f8c60af68d7aca00382ff82865ae795ff3f8ce91.

cbm755 committed:
error fcns: admit a variable of integration to as_integral method

Default to a Dummy if not specified.  `rewrite('Integral')` will always
use the Dummy.  Adjust tests.  Cheap hack to avoid repeating the same
two methods 10 times, but leave a TODO to think about a better solution.

cbm755 committed:
error fcns: test rewrite('Integral') works with deep=True

Its important that `_eval_rewrite_as_Integral(self, *args)` uses `*args*`
and not self when doing the rewriting.  The difference is important for
composition (where `self.args` will be `z` whereas `*args` may contain
a written child from the composition).  One or more of my previous drafts
had this wrong, so add a test which protects this case.

cbm755 commented:
Thank you, I've learned a lot from your input on this; I'm happier with it now.

cbm755 commented on sympy/functions/special/error_functions.py:
```diff
@@ -378,6 +395,14 @@ def _eval_rewrite_as_uppergamma(self, z):
 
+    # TODO: poor style, should inherit/mixin or similar?
+    as_integral = erf.as_integral
+    _eval_rewrite_as_Integral = erf._eval_rewrite_as_Integral
```
This isn't very nice.  But IMHO better than copy-paste the same two methods 10 times...  Conrete suggestions welcome or can leave it for another PR?

cbm755 commented:
Python 2 does not like my `as_integral = erf.as_integral` hack.  *sigh*.

cbm755 committed:
error fcns: workaround python2 by defining functions outside classes

This is probably not worse then the previous hackery.

cbm755 committed:
error fcn tests: only use a dummy for the tests that need it

Most tests pass in a variable of integration; it doesn't need to be
a dummy.  But other tests specifically need a Dummy `t` such that
`str(t) == '_t'`, for those tests, define the Dummy inside the test
itself.

cbm755 commented:
@jksuom can you review again?

jksuom commented:
This looks a bit confusing on the surface, but it seems to work as expected. However, I think that some kind of wrapper could be a better solution. I'll have to look into functools more closely.

jksuom commented:
I may be mistaken but it seems that the following decorator would work:
```
from functools import wraps  # This will set some attributes like __name__ in the wrapped function

def set_variable(func):
    @wraps(func)
    def integral_wrapper(self, t=None):
        if t is None:
            t = Dummy('t')
        return func(self, t)
    return integral_wrapper
```
Then `as_integral = as_integral` could be omitted and `def _as_integral` replaced with
```
    @set_variable
    def as_integral(self, t):
        from sympy import Integral
        return S(2)/sqrt(pi)*Integral(exp(-t**2), (t, 0, self.args[0]))
```

cbm755 commented:
Ok, I've looked into it; nice to learn about decorators, they seem powerful!

I'm not entirely happy with the results: I've put up a new PR for discussion at #12160, maybe you can see a way to improve it?    I will make some comments over there.

Alternatively, I tried to do a "Mixin class" in #12161.  I think I'm more happy with this, but perhaps I've just missed the potential of decorators.
